src/Common/src/System/Numerics/Hashing/Hash.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
+    {
+        public static int Combine(int h1, int h2)
+        {
+            // The jit optimizes this to use the ROL instruction on x86
+            // Related GitHub pull request: dotnet/coreclr#1830
+            uint shift5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
+            return ((int)shift5 + h1) ^ h2;
```

Comment:
> I can see the value in putting this one function in a shared location, but why all of the overloads that take 3-N arguments?

Classes that have 3+ fields, e.g. could be useful [here](https://github.com/dotnet/corefx/pull/8163/files#diff-b6b7e77633e5d7410bd990dd90124e3fR742) in `Cookie`. I don't expect the other methods that aren't used will contribute to much of the code size of the outputted IL; if you want I can have maybe up to 3 overloads of `Combine` in the shared file, and declare the rest in `ValueTuple`.

> do we know that it's a net-gain performance-wise? It may help to reduce collisions, but it would also seem to be a more complicated/expensive operation

It isn't faster than the previous implementation, but I certainly don't think it would be much slower. If I remember correctly the number of registers/instructions used in the disassembly was the same-- before it was shl + add + xor, now it's rol + add + xor. If there was any slowdown, I think the overhead of all the virtual method calls to `EqualityComparer<T>.Default.GetHashCode` would overshadow that here.